source/common/http/conn_manager_impl.cc

```diff
@@ -876,18 +823,20 @@ ConnectionManagerImpl::ActiveStream::ActiveStream(ConnectionManagerImpl& connect
   filter_manager_.streamInfo().setStreamIdProvider(
       std::make_shared<HttpStreamIdProviderImpl>(*this));
 
+  // TODO(chaoqin-li1123): can this be moved to the on demand filter?
+  static const std::string route_factory = "envoy.route_config_update_requester.default";
+  auto factory =
+      Envoy::Config::Utility::getFactoryByName<RouteConfigUpdateRequesterFactory>(route_factory);
   if (connection_manager_.config_.isRoutable() &&
-      connection_manager.config_.routeConfigProvider() != nullptr) {
+      connection_manager.config_.routeConfigProvider() != nullptr && factory) {
```

Comment:
IIUC, if the factory doesn't exist, the ENVOY_BUG will only be called when `requestRouteConfigUpdate` is executed. I think that it should be flagged here and in the "else-if" below.
The reason is that this may be a latent issue that will only occur under a certain condition (route is missing) and may be harder to detect.
I would also consider adding the validation to the route creation if possible, and see if we can avoid surprises on the data-path.

Reply:
yeah I tried to do that,but it turns out just about every config meets the above criterion but most tests (envoy and envoy mobile) don't hit the envoy bug.  I basically think we were lazy on creation an said "always create one or the other" rather than "only create if needed" and I am worried if I try to untagle and actually create when needed I'll break someone's build by missing a corner case.

Reply:
Hmm... strange.
So if we do the following validation:
```
  if (connection_manager_.config_.isRoutable() &&
      connection_manager.config_.routeConfigProvider() != nullptr) {
    ENVOY_BUG(factory, ...); // as defined below.
    if (factory) {
      route_config_update_requester_ =
        std::make_unique<ConnectionManagerImpl::RdsRouteConfigUpdateRequester>(
            connection_manager.config_.routeConfigProvider(), *this);
    }
  } else if (connection_manager_.config_.isRoutable() &&
             connection_manager.config_.scopedRouteConfigProvider() != nullptr &&
             connection_manager.config_.scopeKeyBuilder().has_value()) {
    ENVOY_BUG(factory, ...); // as defined below.
    if (factory) {
      route_config_update_requester_ =
        std::make_unique<ConnectionManagerImpl::RdsRouteConfigUpdateRequester>(
            connection_manager.config_.scopedRouteConfigProvider(),
            connection_manager.config_.scopeKeyBuilder(), *this);
     }
  }
```
there will be false-positives (cases where a route config provider is set, but it is not exercised later - as in `requestRouteConfigUpdate` will not be called).

Very strange, and could probably open up bugs later on, but not sure if there's a better alternative other than the current suggestion.